src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/dynamicProgramming/WrappingDPAlgorithm.java

```diff
@@ -0,0 +1,63 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.dynamicProgramming;
+
+import se.liu.ida.hefquin.engine.federation.FederationAccessException;
+import se.liu.ida.hefquin.engine.queryplan.LogicalOperator;
+import se.liu.ida.hefquin.engine.queryplan.LogicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpMultiwayJoin;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class WrappingDPAlgorithm {
+    protected final QueryOptimizationContext ctxt;
+
+    public WrappingDPAlgorithm( final QueryOptimizationContext ctxt ) {
+        assert ctxt != null;
+        this.ctxt = ctxt;
+    }
+
+    public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws FederationAccessException, QueryOptimizationException {
+        final boolean keepMultiwayJoins = true;
+        final PhysicalPlan initialPhysicalPlan = ctxt.getLogicalToPhysicalPlanConverter().convert(initialPlan, keepMultiwayJoins);
+
+        return optimizePhysicalPlan(initialPhysicalPlan);
+    }
+
+    public PhysicalPlan optimizePhysicalPlan( final PhysicalPlan pp) throws FederationAccessException, QueryOptimizationException {
+        final PhysicalOperatorForLogicalOperator pop = (PhysicalOperatorForLogicalOperator) pp.getRootOperator();
+        final LogicalOperator lop = pop.getLogicalOperator();
+        if ( lop instanceof LogicalOpMultiwayJoin) {
+            final List<PhysicalPlan> children = findChildren(pp);
+            return rewrite( children );
+        }
+        else {
+            return pp;
+        }
+    }
+
+    public List<PhysicalPlan> findChildren( final PhysicalPlan pp ) throws FederationAccessException, QueryOptimizationException {
+        final List<PhysicalPlan> children = new ArrayList<PhysicalPlan>();
+        final int numChildren = pp.numberOfSubPlans();
+        if ( numChildren > 0 ) {
```

Comment:
This if test is not needed. Your for loop handles the case of `numChildren==0` correctly.